p2p/src/disconnection_reason.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Display;
+
+use common::{chain::config::MagicBytes, primitives::time::Time};
+
+use p2p_types::services::Services;
+
+use crate::{
+    error::{ConnectionValidationError, P2pError},
+    protocol::MIN_SUPPORTED_PROTOCOL_VERSION,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DisconnectionReason {
+    AddressBanned,
+    AddressDiscouraged,
+    PeerEvicted,
+    FeelerConnection,
+    ConnectionFromSelf,
+    ManualDisconnect,
+    PingIgnored,
+    SyncRequestsIgnored,
+    TooManyInboundPeersAndThisOneIsDiscouraged,
+    TooManyInboundPeersAndCannotEvictAnyone,
+    UnsupportedProtocol,
+    TimeDiff {
+        remote_time: Time,
+        accepted_peer_time: std::ops::RangeInclusive<Time>,
+    },
+    DifferentNetwork {
+        our_network: MagicBytes,
+    },
+    NoCommonServices,
+    InsufficientServices {
+        needed_services: Services,
+    },
+}
```

Comment:
FYI: here, ConnectionFromSelf and DifferentNetwork are kind of useless, they exist mostly for completeness:
1) It makes no sense to send WillDisconnect for ConnectionFromSelf if the connection is indeed a self-connection. But technically, detection of self-connections may have false positives (if 2 randomly chosen 64-bit nonces happen to be the same), so I decided to do it.
2) WillDisconnect for DifferentNetwork will never be received by a peer in reality, because both sides perform the check "simultaneously" after the handshake and will disconnect each other before any other message will have a chance to be handled. 

UnsupportedProtocol is also useless at this moment; it'll only become useful after/if we drop the support for V3.